src/extensions/extension-store.ts

```
@@ -0,0 +1,174 @@
+        let instance = this.getById(manifestPath);
+        if (!instance) {
+          const LensExtensionClass = extensionModule.default;
+          instance = new LensExtensionClass({ ...manifest, manifestPath, id: manifestPath }, manifest);
+          instance.enable(getLensRuntimeEnv());
+          this.extensions.set(manifestPath, instance); // save
+        }
+      })
+    }, {
+      fireImmediately: true,
+      delay: delay,
+    })
+  }
+
+  getExtensionByManifest(manifestPath: string): InstalledExtension {
+    let manifestJson: ExtensionManifest;
+    let mainJs: string;
+    try {
+      manifestJson = __non_webpack_require__(manifestPath); // "__non_webpack_require__" converts to native node's require()-call
+      mainJs = path.resolve(path.dirname(manifestPath), manifestJson.main);
+      mainJs = mainJs.replace(/\.ts$/i, ".js"); // todo: compile *.ts on the fly?
+      const extensionModule = __non_webpack_require__(mainJs);
+      return {
+        manifestPath: manifestPath,
+        manifest: manifestJson,
+        extensionModule: extensionModule,
+      }
+    } catch (err) {
+      console.error(`[EXTENSION-STORE]: can't load extension at ${manifestPath}: ${err}`, { manifestJson, mainJs });
+    }
+  }
+
+  @action
+  async loadExtensions() {
+    const extensions = await this.loadFromFolder(this.folderPath);
+    const extManifestMap = new Map(extensions.map(ext => [ext.manifestPath, ext]));
+    this.installed.replace(extManifestMap);
+  }
+
+  async loadFromFolder(folderPath: string): Promise<InstalledExtension[]> {
+    const paths = await fs.readdir(folderPath);
+    const manifestsLoading = paths.map(fileName => {
+      const absPath = path.resolve(folderPath, fileName);
+      const manifestPath = path.resolve(absPath, "package.json");
+      return fs.access(manifestPath, fs.constants.F_OK)
+        .then(() => this.getExtensionByManifest(manifestPath))
+        .catch(() => null)
```

Why do we ignore the error, shouldn't we at least debug log it?

---

Yes, to extension developer it would be very beneficial information. I assume this is still a moving target, so I'll create one more todo item to epic issue about error handling.